### PR TITLE
Pgtyped orguser audit

### LINKED ISSUE #4081: OrganizationUser audit trail
## Issue - Enhancement

One day, it'd be nice to have an `OrganizationUserJournal` table that is updated each time something important happens to an Org User. Today, I think this should be when a user is created, marked active, marked inactive, or removed from an organization.

Today we can create a plot from backend data of the number of active users over time:

image

What I really want to be able to do is create analyses for ourselves and our users of how their organization's activity has changed over time. Because we only store the current state, this kind of historical analysis is currently impossible.

### Acceptance Criteria (optional)
   - Migration to add `OrganizationUserAudit` table in POSTGRES
   - Row schema designed that include 'eventDate' timestamp
   - Rows added when a user is created, marked active, marked inactive, or removed from an organization

- **Estimated effort:** X hours ([see CONTRIBUTING.md](https://github.com/ParabolInc/parabol/blob/master/CONTRIBUTING.md#points-and-sizes))

### PR DESCRIPTION
**Pull request by tiffanyhan**
Resolve #4081

**AC:**
Audit row(s) added when:
- [ ] a new user and org is bootstrapped
- [ ] a new org is created
- [ ]  a user joins a new org
- [ ] a user leaves an org
- [ ] a user is paused
- [ ] a user in unpaused
- [ ] a user deletes their account, thereby leaving all joined organizations

### CODE REVIEW

**Comment by tiffanyhan** [comment]
Working with pgtyped was pretty delightful. i especially appreciated its enablement of named parameters and parameter expansions. Next steps would be:

- [ ] adding pgtyped watch mode to `dev.js`
- [ ] creating wrapper functions to manage pg pool connections and transactions. so we don't forget to release the client back to the pool, etc. this is in a lot of other libs that do a lot of other things, but we can probably just right it ourself.

File: packages/server/postgres/queries/OrgUserAudit.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
+/** Types generated for queries found in "packages/server/postgres/queries/OrgUserAudit.sql" */
```
Comment (mattkrick) [inline packages/server/postgres/queries/OrgUserAudit.ts:1]: can we keep generated files in their own directory? it's always nice to be able to easily blow away a generated directory when something goes wrong or exclude from git if we need to down the road

File: package.json [hunk 2]
```diff
@@ -24,6 +24,7 @@
     "start": "pm2-runtime start pm2.config.js",
     "heroku-postbuild": "echo 'Skipping build'",
     "build": "node scripts/prod.js",
+    "pg:build": "pgtyped -c ./packages/server/postgres/pgtypedConfig.js",
```
Comment (mattkrick) [inline package.json:27]: what do you recommend as a pattern for calling this? should we call it on ever server restart?
perhaps we could use pm2 to run a dedicated instance that watches a directory & runs it whenever that dir changes?
we can leave that piece out of scope for this PR, but it'll be good to think about for a future PR

File: packages/server/postgres/getPgClient.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import getPgPool from './getPgPool'
```
[the remark this answers is not in the document]
Comment (tiffanyhan) [reply packages/server/postgres/getPgClient.ts:1]: i see the two as distinct. `getPgPool` gives access to the pool. you can then call `pool.query` which will get any available client from the pool then automatically release it after you call a single query.

`pool.connect` returns a single client which you can then use to do multiple queries. useful for transactions or when you need to ensure multiple queries go on the same connection in general. if using this pattern with vanilla `pg`, you have to remember to release the client manually.

`getPgPool` would be useful for usage of the first type, and `getPgClient` would be useful for usage of the second type. 

in `pgTyped`, the API requires passing a client into the generated function, so that's why `getPgClient` was used in this instance. i kept the two getters as separate as we're setting up the postgres API in general, but it could be that we only end up using the client getter in practice.

File: packages/server/postgres/migrations/1611342655964_addOrganizationUserAuditTable.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+export const shorthands: ColumnDefinitions | undefined = undefined;
+
+export async function up(pgm: MigrationBuilder): Promise<void> {
+  pgm.sql(`
+    CREATE TABLE "OrganizationUserAudit" (
+      id SERIAL PRIMARY KEY,
+      "orgId" VARCHAR(100) NOT NULL,
+      "userId" VARCHAR(100) NOT NULL,
+      "eventDate" TIMESTAMP WITH TIME ZONE NOT NULL,
+      "eventType" INT NOT NULL
```
[the remark this answers is not in the document]
Comment (tiffanyhan) [reply packages/server/postgres/migrations/1611342655964_addOrganizationUserAuditTable.ts:13]: yea that is definitely a downside to keeping enums in the app. 

i think other pros of enums besides saving space is its easier for devs to change the enum without creating a migration, and it keeps business logic separate from data.

> in a future where we might have a data analytics team that works off of our DB, what's the recommended pattern for them to match the numeric value to the typescript definition?

that's a great question. i'd expect anyone working on data to be able to find the mapping in the codebase, but that may or may not be a reasonable expectation. i suppose there could also be some hook that updates the enum definitions to a wiki, so there's still a single source of truth in addition to an easy place for them to find it.

File: packages/server/postgres/types/OrgUserAuditEventTypeEnum.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+enum OrgUserAuditEventTypeEnum {
+  added = 1,
+  activated,
```
[the remark this answers is not in the document]
Comment (tiffanyhan) [reply packages/server/postgres/types/OrgUserAuditEventTypeEnum.ts:3]: that's a great point. what did you have specifically in mind by, being "explicit with our namings"?

File: packages/server/postgres/queries/OrgUserAudit.ts (hunk 1 above)
Comment (tiffanyhan) [reply packages/server/postgres/queries/OrgUserAudit.ts:1]: yeah sure!

File: package.json (hunk 2 above)
Comment (tiffanyhan) [reply package.json:27]: pgtyped cli has watch mode, and i was thinking we could call this from `dev.js`. disclaimer: i haven't tried invoking watch mode manually yet, so i can't yet vouch for its reliability.

https://pgtyped.now.sh/docs/cli

File: packages/server/postgres/migrations/1611342655964_addOrganizationUserAuditTable.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (tiffanyhan) [reply packages/server/postgres/migrations/1611342655964_addOrganizationUserAuditTable.ts:13]: just to be fair, other pros of keeping enums in the db are we get more semantic meaning when querying the db directly plus database-level validation of enums.

File: packages/server/postgres/helpers/OrgUserAudit.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  orgIds: string[],
+  userId: string,
+  eventType: OrgUserAuditEventTypeEnum,
+  eventDate: Date = new Date()
+) => {
+  const pgClient = await getPgClient()
+  const auditRows = orgIds.map((orgId) => ({orgId, userId, eventType, eventDate}))
+  const parameters = {auditRows}
+  await insertQuery.run(parameters, pgClient)
+  pgClient.release()
```
[the remark this answers is not in the document]
Comment (tiffanyhan) [reply packages/server/postgres/helpers/OrgUserAudit.ts:15]: i'm not sure i understand the question. what would be the purpose of having a pool of available connections if the client is a singleton?

File: packages/server/postgres/helpers/OrgUserAudit.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import {insertOrgUserAudit as insertQuery} from '../queries/OrgUserAudit'
```
Comment (mattkrick) [inline packages/server/postgres/helpers/OrgUserAudit.ts:1]: can we keep the file name == function name & export it as the default? doing this means we don't have to rename generated query when we import it

File: packages/server/postgres/getPgClient.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattkrick) [reply packages/server/postgres/getPgClient.ts:1]: that'd be a dealbreaker for pgtyped if true!
can you show me where you're seeing it requires a client & not a pool? from what i see, the param def looks like:
```ts
export interface IDatabaseConnection {
    query: (query: string, bindings: any[]) => Promise<{
        rows: any[];
    }>;
}
```

lets get rid of `getPgClient` all together, it really only does 1 thing & that 1 thing is dangerous (provide an unterminating connection). when the time for transactions comes, those transactions can grab a connection in their own functions.

File: packages/server/postgres/helpers/OrgUserAudit.ts (hunk 7 above)
Comment (tiffanyhan) [reply packages/server/postgres/helpers/OrgUserAudit.ts:1]: @mattkrick hmm yeah that would be nice. if i'm not mistaken, i don't think there's any way to configure pgtyped to export a generated fn as the default. i think the intention of the author was for each sql file to contain multiple queries pertaining to one table.

File: packages/server/postgres/getPgClient.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (tiffanyhan) [reply packages/server/postgres/getPgClient.ts:1]: yeah you're right. you can pass in either a pool or client. i was just going by the docs instead of looking at the types and source code.

[the remark this answers is not in the document]
Comment (tiffanyhan) [reply packages/server/postgres/getPgClient.ts:1]: > lets get rid of getPgClient all together, it really only does 1 thing & that 1 thing is dangerous (provide an unterminating connection). when the time for transactions comes, those transactions can grab a connection in their own functions.

i was seeing the client getter as more placeholder for work in progress. that's a good point though, somebody looking at the code as-is could be encouraged to do bad behavior.

[the remark this answers is not in the document]
Comment (tiffanyhan) [reply packages/server/postgres/getPgClient.ts:1]: client getter deleted

File: packages/server/postgres/migrations/1611342655964_addOrganizationUserAuditTable.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattkrick) [reply packages/server/postgres/migrations/1611342655964_addOrganizationUserAuditTable.ts:13]: > i think other pros of enums besides saving space is its easier for devs to change the enum without creating a migration, and it keeps business logic separate from data.

if the type was a varchar would a migration still be needed?

> just to be fair, other pros of keeping enums in the db are we get more semantic meaning when querying the db directly plus database-level validation of enums.

could you tell me more about the semantic meaning?
how does the DB validate enums? is there something preventing me from passing in `249` when that enum has no reference?

File: packages/server/postgres/helpers/OrgUserAudit.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import {insertOrgUserAudit as insertQuery} from '../queries/generated/OrgUserAudit'
+import getPgPool from '../getPgPool'
+import OrgUserAuditEventTypeEnum from '../types/OrgUserAuditEventTypeEnum'
+
+export const insertOrgUserAudit = async (
```
Comment (mattkrick) [inline packages/server/postgres/helpers/OrgUserAudit.ts:5]: can we rename this file to `insertOrgUserAudit`?

File: packages/server/postgres/queries/src/OrgUserAudit.sql [hunk 9]
```diff
@@ -0,0 +1,10 @@
+/* 
+  @name insertOrgUserAudit 
+  @param auditRows -> ((orgId, userId, eventDate, eventType)...)
+*/
+INSERT INTO "OrganizationUserAudit" (
```
Comment (mattkrick) [inline packages/server/postgres/queries/src/OrgUserAudit.sql:5]: let's rename this to `insertOrgUserAuditRow`

File: packages/server/postgres/helpers/OrgUserAudit.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
+import {insertOrgUserAudit as insertQuery} from '../queries/generated/OrgUserAudit'
```
Comment (mattkrick) [inline packages/server/postgres/helpers/OrgUserAudit.ts:1]: if we change the sql file name to `insertOrgUserAuditRow` then we can import that without having to rename it up here

File: packages/server/postgres/queries/src/OrgUserAudit.sql [hunk 11]
```diff
@@ -0,0 +1,10 @@
+/* 
+  @name insertOrgUserAudit
```
Comment (mattkrick) [inline packages/server/postgres/queries/src/OrgUserAudit.sql:2]: lets have the name match the filename: insertOrgUserAuditRow

File: packages/server/postgres/getPgPool.ts [hunk 12]
```diff
@@ -2,7 +2,7 @@ import {Pool} from 'pg'
 import PROD from '../PROD'
 
 let pool: Pool
-const getPG = () => {
+const getPgPool = () => {
```
Comment (mattkrick) [inline packages/server/postgres/getPgPool.ts:5]: lets rename to getPg since we'll be using this a lot!

File: packages/server/postgres/migrations/1611342655964_addOrganizationUserAuditTable.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattkrick) [reply packages/server/postgres/migrations/1611342655964_addOrganizationUserAuditTable.ts:13]: after chatting, let's use the postgres enum type & use pg-typed to generate the enum in typescript land. then, we'll get checking for free in typescript PLUS validation at the DB level. win win!

File: packages/server/postgres/migrations/1611342655964_addOrganizationUserAuditTable.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+export const shorthands: ColumnDefinitions | undefined = undefined;
+
+export async function up(pgm: MigrationBuilder): Promise<void> {
+  pgm.sql(`
+    CREATE TABLE "OrganizationUserAudit" (
+      id SERIAL PRIMARY KEY,
+      "orgId" VARCHAR(100) NOT NULL,
+      "userId" VARCHAR(100) NOT NULL,
+      "eventDate" TIMESTAMP WITH TIME ZONE NOT NULL,
```
Comment (mattkrick) [inline packages/server/postgres/migrations/1611342655964_addOrganizationUserAuditTable.ts:12]: no timezone needed here, our DB & servers are all in UTC & if we need to convert it to a specific TZ we can, but we don't need to know the source tz.

File: packages/server/postgres/queries/generated/insertOrgUserAuditQuery.ts [hunk 14]
```diff
@@ -0,0 +1,40 @@
+/** Types generated for queries found in "packages/server/postgres/queries/src/insertOrgUserAuditQuery.sql" */
+import { PreparedQuery } from '@pgtyped/query';
+
+export type OrganizationUserAuditEventTypeEnum = 'added' | 'activated' | 'inactivated' | 'removed';
```
Comment (mattkrick) [inline packages/server/postgres/queries/generated/insertOrgUserAuditQuery.ts:4]: :fire:

File: packages/server/graphql/mutations/helpers/createTeamAndLeader.ts [hunk 15]
```diff
@@ -69,4 +70,5 @@ export default async function createTeamAndLeader(userId: string, newTeam: Valid
   if (!organizationUser) {
     await adjustUserCount(userId, orgId, InvoiceItemType.ADD_USER)
   }
+  await insertOrgUserAudit([orgId], userId, 'added')
```
Comment (mattkrick) [inline packages/server/graphql/mutations/helpers/createTeamAndLeader.ts:73]: why is this necessary? isn't this handled already in `adjustUserCount`? will calling it here create a dupe?

Comment (tiffanyhan) [reply packages/server/graphql/mutations/helpers/createTeamAndLeader.ts:73]: it's not handled in `adjustUserCount` bc an `OrganizationUser` is inserted separately on line 44. Can we get rid of that insertion in the string of promises, and just call `adjustUserCount` to make it consistent?

Comment (mattkrick) [reply packages/server/graphql/mutations/helpers/createTeamAndLeader.ts:73]: shoot i'm not seeing it. can you link to the link where `OrganizationUser` is getting inserted on L44?

Comment (tiffanyhan) [reply packages/server/graphql/mutations/helpers/createTeamAndLeader.ts:73]: oops sorry, on line 44 it just fetches to check if an `OrganizationUser` was already made. `adjustUserCount` never gets called bc the check always finds an `OrganizationUser` already. 

it looks like `bootstrapNewUser` and `addOrg` both call `createOrg` which creates an `Organization` as well as `OrganizationUser` before calling `createTeamAndLeader`.

Comment (tiffanyhan) [reply packages/server/graphql/mutations/helpers/createTeamAndLeader.ts:73]: maybe we can change `createNewOrg` to only create an `Organization` and not an `OrganizationUser`? it looks like `createNewOrg` is only called by `bootstrapNewUser` and `addOrg` and nothing else, so should be a safe change.

Comment (mattkrick) [reply packages/server/graphql/mutations/helpers/createTeamAndLeader.ts:73]: ah, makes sense.
we don't want to do org user audit things in `createTeamAndLeader` because that can be called by `addTeam`, which would result in multiple records.
instead, let's
- remove L70-L73 here because that conditional branch is never true
- add the `insertOrgUserAudit` call to `createNewOrg`

File: packages/server/graphql/mutations/helpers/createTeamAndLeader.ts [hunk 16]
```diff
@@ -38,7 +36,7 @@ export default async function createTeamAndLeader(userId: string, newTeam: Valid
     orgId
   })
 
-  const [organizationUser] = await Promise.all([
+  await Promise.all([
     r
```
[the remark this answers is not in the document]
Comment (tiffanyhan) [reply packages/server/graphql/mutations/helpers/createTeamAndLeader.ts:40]: oh true true

**Comment by tiffanyhan** [comment]
https://github.com/ParabolInc/parabol/issues/4554#issuecomment-770314089

^^ i'm beginning to have second thoughts about pgtyped, gonna try and make an alt PR for this with zapatos tmrw.

**Comment by mattkrick** [comment]
for updating a field to null it should be like
@tiffanyhan 
```sql
UPDATE foo
SET bar = NULL
WHERE baz = 'hi';
```
